### ISSUE TITLE
P/Invoke from .Net Core in WSL calls DLLMain

### ISSUE THREAD

**Issue by dcorbett-fivebridgesllc** [dotnet/coreclr#25142]
I am porting an application that works in .net framework to .net core and Linux.
It uses DLLImport to cause P/Invoke to be called on the DLL or .so in Linux.
Under WSL when I call the library function, it is always attempting to call DLLMain, and not finding it.
I can't build DLLMain in Linux C++ as I can in Windows C++ because "far" isn't supported, thus it will never be able to find DLLMain.
I've tried:
[DllImport(@"libAvmModel.so", CharSet = CharSet.Ansi, CallingConvention = CallingConvention.StdCall)]
public static extern void GetAVMvalue(…);

[DllImport(@"libAvmModel.so")]
public static extern void GetAVMvalue(…);

After setting LD_DEBUG=libs
I see the entry point it's trying:
6157: /mnt/c/dev/avm/Avm.ServiceHost/bin/x64/Debug/netcoreapp2.2/publish/libAvmModel.so: error: symbol lookup error: undefined symbol: DllMain (fatal)
Target Property: 0x7fb0f8012850 OrderId: 27e9581d-552e-e811-80f4-44a8423692d6-01/01/2019 .
Segmentation fault (core dumped)

Alternatively when I build the code in docker, and run it in a container, I am now getting:

calling init: /app/libAvmModel.so
/app/libAvmModel.so: error: symbol lookup error: undefined symbol: PAL_RegisterModule (fatal)

**Comment by dcorbett-fivebridgesllc** [comment]
FYI - My WSL was running Ubuntu 18.04
Running .net core 2.2
C++ is compiled with G++ 7.4 using CMake in WSL

**Comment by jkoritzinsky** [comment]
The DllMain on a native library on non-Windows should be optional. If you don't have one, everything should work fine. Same with the PAL_RegisterModule. Are you seeing a crash with it not finding the DllMain or PAL_RegisterModule entrypoints?

**Comment by jkotas** [comment]
Also, the .NET Core 3.0 does not even try to look for DllMain anymore. You can give this a try on .NET Core 3.0.

**Comment by dcorbett-fivebridgesllc** [comment]
I am aware it should be optimal.    Yes, I am seeing a crash.   The only error is one of those, and they come when I have LD_DEBUG=libs

**Comment by jkoritzinsky** [comment]
Are you able to run your application under a debugger or open the coredump in a debugger and see where the segfault is happening? I'd be curious to see where in the runtime we're getting a segfault on that codepath.

**Comment by dcorbett-fivebridgesllc** [comment]
No.   Haven't seen the coredump, and couldn't get the debugger to work in WSL.
Haven't yet tried in Docker.

**Comment by jkoritzinsky** [comment]
lldb doesn't work in WSL1. gdb should work though.

**Comment by dcorbett-fivebridgesllc** [comment]
I did try .net core 3.0 in my docker work ( have moved passed WSL since I originally posted this to the Developer Community and then was redirected here ).    It didn't mention Dllmain OR PAL_RegisterModule but still crashed on me around the same place.. ;-(

I've used gdb in prior lives with C/C++ but not with dotnet core.   What's the best guide to using that in Linux/Docker?    The setup seems significant from what I can find.

**Comment by jkotas** [comment]
> what's the best guide to using that in Linux/Docker?

You need to run the container with `--cap-add=SYS_PTRACE --security-opt seccomp=unconfined` to make the debugger work. Other than that, diagnosing this crash should be not different from how you would diagnose crash like this for C/C++.

**Comment by dcorbett-fivebridgesllc** [comment]
Getting it to crash is the problem..  It just exits ( with 0 ), at the point the P/Invoke should be happening.  It's obviously loading the .so as it's calling init on it and all referenced .so.   Right now, the only idea I have is to debug the coreCLR to load the managed code and hope that I can set a breakpoint in the managed code right before the P/Invoke...   Or perhaps the CoreCLR implantation of DLLAttribute??

**Comment by dcorbett-fivebridgesllc** [comment]
Thanks everyone, in Ubuntu, they have disabled the crash support by default -- I found help though at this wiki page, and was able to turn it on:   https://wiki.ubuntu.com/Apport 

Once the crash happened and I looked at a core dump, I realized that it was a memory issue caused by an error in the application ( that I had diagnosed previously ) and was unrelated to issues I mentioned above.

We are passing a struct by value into the C++ function that is expecting a pointer.   In Windows, this was automatically handled, as if it as passed by reference.  In Linux it behaves differently ( but not incorrectly ).   CHanging the calling structure to use ref fixed the issue.
